src/MizarData/include/MizarData/MizarPairedData.h

```diff
@@ -198,7 +203,45 @@ class MizarPairedDataTmpl {
   absl::flat_hash_map<TID, uint64_t> tid_to_callstack_samples_counts_;
 };
 
-using MizarPairedData = MizarPairedDataTmpl<MizarDataProvider, FrameTrackManager>;
+template <typename Data, auto GetSamplingIntervalsNs>
+uint64_t RobustSamplingPeriodEstimate(const absl::flat_hash_set<orbit_mizar_base::TID>& tids,
+                                      uint64_t min_timestamp, uint64_t max_timestamp,
+                                      const Data& data) {
+  std::vector<uint64_t> intervals = GetSamplingIntervalsNs(
+      tids, min_timestamp, max_timestamp, data.GetCaptureData().GetCallstackData());
+  absl::c_sort(intervals);
+
+  constexpr double kOutlierRate = 0.025;
+  const size_t cutoff_index = static_cast<size_t>((1 - kOutlierRate) * intervals.size());
+
+  if (cutoff_index + 1 > intervals.size()) {
+    return data.GetNominalSamplingPeriodNs();
+  }
+  const uint64_t cutoff = intervals[cutoff_index];
+
+  struct TotalAndCount {
+    uint64_t total{};
+    size_t count{};
+
+    // round(total / count)
+    uint64_t ComputeAverage() { return (total + (count / 2)) / count; }
```

Comment:
Can you explain why "total + (count/2) / count" and why this is round(total/count)? I think I am missing something crucial here.

Reply:
Discussed offline.

Reply:
This works. Basically you are adding ~0.5 and making a floor using integers. For count being even is exactly 0.5 and for count being odd is a bit less but since you cannot have the fraction 0.5 works anyway.